### PR TITLE
5872-workingDirectory-badly-resolved-if-its-path-contains-non-ascii-characters

### LINKED ISSUE #5872: workingDirectory badly resolved if its path contains non-ascii characters
To reproduce:
1. create a folder with a name containing a non-ascii character: `mkdir loïc && cd loïc`
1. get a fresh pharo image and vm: `curl get.pharo.org/64/80+vm |  bash`
1. `./pharo-ui Pharo.image eval "(FileLocator workingDirectory) asFileReference exists"` => false
1. `./pharo-ui Pharo.image eval "(FileLocator workingDirectory) asFileReference fullName"` => /private/tmp/loÃ¯c

It is because the primitive `primitiveGetCurrentWorkingDirectory` just returns a raw byte array that should be decoded with the proper Character encoding.

Running `ZnCharacterEncoder utf8 asZnCharacterEncoder decodeBytes: OSPlatform current currentWorkingDirectoryPath asByteArray`gives the expected result: /private/tmp/loïc

### PR DESCRIPTION
**Pull request by demarey**
Fixes #5872 workingDirectory badly resolved if its path contains non-ascii characters

### CODE REVIEW

**Comment by tesonep** [comment]
@demarey check that the modified method relies in a method that is loaded in a package that is loaded later. As this is core part of the system, we need to be aware that maybe #environment is not accessible.

**Comment by Ducasse** [comment]
Christophe I let you have a look.

File: src/System-Platforms/OSPlatform.class.st [hunk 1]
```diff
@@ -97,26 +97,34 @@ OSPlatform >> current [
...
 	^ self currentWorkingDirectoryPathWithBuffer: buffer
 ]
 
+{ #category : #accessing }
+OSPlatform >> currentWorkingDirectoryRawPath [
+	"This method calls the method getPwdViaFFI with arugement of a buffer size. 
+	By default it uses the defaultMaximumPathLength of each subclass as the buffer size.
+	We return a raw byte array that should be decoded with the proper encoding."
+	^ (self currentWorkingDirectoryPathWithBufferSize: self defaultMaximumPathLength) asByteArray
+
```
Comment (guillep) [inline src/System-Platforms/OSPlatform.class.st:125]: Is the `asByteArray` needed?

File: src/System-Platforms/OSPlatform.class.st [hunk 2]
```diff
@@ -97,26 +97,34 @@ OSPlatform >> current [
 
 { #category : #accessing }
 OSPlatform >> currentWorkingDirectoryPath [
-	"This method calls the method getPwdViaFFI with arugement of a buffer size. By default it uses the defaultMaximumPathLength of each subclass as the buffer size."
-	^ self currentWorkingDirectoryPathWithBufferSize: self defaultMaximumPathLength
-
+	"We MUST decode raw bytes to support non-Ascii characters"
+	^ self currentWorkingDirectoryRawPath decodeWith: self environment defaultEncoding
```
Comment (guillep) [inline src/System-Platforms/OSPlatform.class.st]: As @tesonep says, `environment` is an extension method, so this method is creating a circular dependency. (System-Platforms <-> System-OSEnvironments)

**Review by guillep — CHANGES_REQUESTED** [review]

**Comment by guillep** [comment]
@demarey could you give us an idea of when you'll be checking this?

**Comment by demarey** [comment]
@guillep 
OSPlatform>>#currentWorkingDirectoryPath uses 
- #decodeWith: packaged in Zinc-Character-Encoding-Core (loaded in Bootstrap kernel group) => ok
- #defaultEncoding packaged in System-OSEnvironments. The problem here is that System-OSEnvironments package is loaded with UFFI because it uses a lot UFFI. BaselineOfUnifiedFFI is loaded either by MorphicCore or Athens (loaded by BaselineOfIDE).
I have the feeling that we should move the #defaultEncoding definition from OSEnvironment to OSPlatfrom. WDYT?

**Comment by guillep** [comment]
> * #defaultEncoding packaged in System-OSEnvironments. The problem here is that System-OSEnvironments package is loaded with UFFI because it uses a lot UFFI. BaselineOfUnifiedFFI is loaded either by MorphicCore or Athens (loaded by BaselineOfIDE).
>   I have the feeling that we should move the #defaultEncoding definition from OSEnvironment to OSPlatfrom. WDYT?

Looks good enough to me. Can we try that?

**Comment by tesonep** [comment]
Really cool analysis, it will be nice to extend the dependency analyser to
know where the packages are loaded and its dependencies.

What do you think?

**Comment by tesonep** [comment]
And also, I share it should be a good solution to move the method

**Comment by guillep** [comment]
Thanks! Let's wait for the CI.

**Review by guillep — APPROVED** [review]
Good! tests are green!

**Pull request merged**